### ISSUE TITLE
[request] multi-line string where indentation get's eaten

### ISSUE THREAD

**Issue by clankill3r** [odin-lang/Odin#1971]
One thing I dislike about multi-line comments is that they can screw with the look of the file due to the indentation.
Here a tiny example (I think it's worse when the multi-line strings covers a lot more lines and contains code (like shader code)):

<img width="694" alt="Screenshot 2022-08-22 at 21 57 31" src="https://user-images.githubusercontent.com/738650/186008611-d46d06b6-651d-4b22-9a26-f8b6bd3160ed.png">

I would much more prefer something like this:

<img width="683" alt="Screenshot 2022-08-22 at 21 56 52" src="https://user-images.githubusercontent.com/738650/186008703-3992c8cd-b303-4f53-96ca-08b8beb12462.png">

But that results in:

<img width="490" alt="Screenshot 2022-08-22 at 21 57 11" src="https://user-images.githubusercontent.com/738650/186008749-120c0fcd-3fd7-4778-993b-b3d235a8ffda.png">

The problem is knowing how much white space to eat.
My idea is to have something like this:

<img width="712" alt="Screenshot 2022-08-22 at 22 38 44" src="https://user-images.githubusercontent.com/738650/186014520-cb808b17-5b73-43b8-8122-7ddcb04d5e54.png">

The idea of how it works:
- `#eat_prefixed_whitespace` has to be present
- the second line has to contain a `|` character
- the position of this character is x tabs in, where x is how many tabs it should eat minus one (cause the second line might contain content as well so we can't put it there
- on the lines after the second line with the `|`, if any other character is found then a tab in this prefixed area, then a error should be thrown

* I typed some ideas for when people use spaces instead of tabs as well, but people that use spaces should be shot...

**Comment by clankill3r** [comment]
bonus
<img width="852" alt="Screenshot 2022-08-22 at 23 16 37" src="https://user-images.githubusercontent.com/738650/186020651-d8fed6be-56a0-4417-8bee-2f96a6c49aba.png">

**Comment by gingerBill** [comment]
Easiest solution is to use Python like triple quote strings.

**Comment by clankill3r** [comment]
yeah that would be easier. Still there needs to be something to indicate how much to eat.

**Comment by z64** [comment]
I've enjoyed using heredocs in languages like Crystal:

```cr
sql = <<-SQL
  SELECT
    a,
    b,
    c
  FROM bar
  SQL

# yields "SELECT\n  a,\n  b,\n  c\nFROM bar\n"
```

The way it works is leading whitespace is aligned with whatever column the terminating marker is at.

For example, this yields exactly the same string:

```cr
sql = <<-SQL
      SELECT
        a,
        b,
        c
      FROM bar
      SQL
```

So it would be nice if it worked the same way, where leading whitespace is trimmed up to the column of the closing fence:

```odin
sql := """
    SELECT
      a,
      b,
      c
    FROM bar
    """
assert(sql == "SELECT\n  a,\n  b,\n  c\nFROM bar\n")
```

---

While I don't expect Odin to adopt heredocs, I will mention another nice feature *specifically* with heredoc syntax, is they can be recognized by frontends like GitHub and use the heredoc identifier (`SQL`) to pick different syntax highlighting for that block of text.

**Comment by gingerBill** [comment]
**IFF** I am going to do this as a feature in the language, it will be Python-like triple quotes as I suggested.

Please don't suggest any other lexical/syntax bits please.

**Comment by z64** [comment]
@gingerBill To be clear, I was not suggesting Odin implement heredocs like Crystal precisely as illustrated.

I was suggesting that I would really like leading space to be trimmed in the same manner, using the closing fence as a guide - so that the last example that I wrote passes the assert.

**Comment by Feoramund** [comment]
This looks like an anti-feature in that it allows you to indent strings all the way out to the starting indent, sacrificing horizontal space. String literals can be concatenated these days with `+` and that works across lines, if this is really an issue.

**Comment by clankill3r** [comment]
This works:

```odin
package main

import "core:fmt"
import "core:strings"
import "base:runtime"


main :: proc() {
    fmt.println(foo(`When using a multi-line
                    string then bla bla.
                    Lorem ipsum.`))
    
}

foo :: proc(text: string,  loc: runtime.Source_Code_Location = #caller_location) -> string {
    fmt.println(loc.column)

    lines := strings.split(text, "\n");
    for i in 1..< len(lines) {
        line := lines[i]
        if len(line) > int(loc.column+3) {
            lines[i] = line[loc.column+3:]
        }
        else {
            lines[i] = ""
        }
    }
    return strings.join(lines, "\n")
}
```
